thunder/dynamo/splitter.py

```diff
@@ -0,0 +1,166 @@
+from collections.abc import Callable
+
+import torch
+from torch.fx.passes.split_module import split_module
+
+from thunder.dynamo.utils import (
+    SubgraphInfo,
+    CompiledFunction,
+    CompilerType,
+    SplitReason,
+    SplitReasonType,
+    is_node_supported_by_thunder,
+    get_nodes_in_unsupported_ctx_regions,
+    update_node_and_submodule,
+    recompile_graph,
+)
+
+
+def _splitter(
+    gm: torch.fx.GraphModule,
+    thunder_jit: Callable,
+    torch_inductor: Callable,
+    _unused_sample_args: list[torch.SymInt, torch.Tensor],
+) -> tuple[torch.fx.GraphModule, SubgraphInfo]:
+    """
+    This method will split graph into multiple graph modules based on thunder supported operations.
+    This function will try to split the graph in contiguous partitions.
+
+    Example:
+        # All operations are supported by thunder
+        class GraphModule(torch.nn.Module):
+            def forward(self, L_x_: "f32[2]"):
+                l_x_ = L_x_
+
+                y: "f32[2]" = torch.sin(l_x_)
+                matmul: "f32[]" = torch.matmul(l_x_, y);  l_x_ = y = None
+                return (matmul,)
+
+        # Split Graph: All operations are supported by thunder, we will see only one partition.
+        class GraphModule(torch.nn.Module):
+            def forward(self, l_x_: "f32[2]"):
+                thunder_1 = self.thunder_1(l_x_);  l_x_ = None
+                return (thunder_1,)
+
+            class thunder_1(torch.nn.Module):
```

Comment:
Some questions not for this PR.

Do we have to represent the call to `thunder_1` as a callable class? Calling a callable class is slower than calling a function.

Reply:
Currently, in splitter we call `thunder.jit` with a split graph module and hence `thunder.jit` returns an OptimizedModule. 

I don't have a lot of idea about the cost of calling a function vs callable but I am wondering if it would be a bottleneck compared to the cost of actual computation?

Thanks!

Reply:
@jjsjann123 had a similar question about Python latencies recently -- how do we quickly get a sense for how much they matter?

fyi @tfogal , too, in case he'd like to share some thoughts

Let's say we have two approaches we're considering, A and B. Let's say that approach A is faster than approach B, and we want to know if switching from B to A is worthwhile. One question I like to think about is how many times would we have to do B before it's 10 microseconds slower than doing A the same number of times? We might not care about 10 microseconds, but that seems like a good proxy for the smallest unit of time we could care about. 

We can test this by running the experiment in Python, of course. Maybe something like:


```
import time

def foo(a, b):
  return

class mycls:

  def __call__(self, a, b):
    return

iterations = int(1e6)

start = time.time_ns()
for _ in range(iterations):
  foo(1, 2)
end = time.time_ns()

elapsed_ns = end - start
fn_elapsed_us = (elapsed_ns / 1000)
print(f"fn call {fn_elapsed_us=}")

x = mycls()
start = time.time_ns()
for _ in range(iterations):
  x(1, 2)
end = time.time_ns()
elapsed_ns = end - start
callable_class_elapsed_us = (elapsed_ns / 1000)
print(f"class call {callable_class_elapsed_us=}")
iters_before_10us_slower = 10 / ((callable_class_elapsed_us - fn_elapsed_us) / iterations)
print(f"calls before it's 10us slower: {iters_before_10us_slower}")
```

On my laptop this reports that we'd have to call the callable class ~1000 times before we'd see a 10us difference vs. calling a function. I'm guessing we expect to call callable classes like this fewer than 1000 times per program, too, so the e2e impact is only likely to be a priority if we are cpu bound and want to improve perf by a few microseconds